athenahealth/lab_results.go

```diff
@@ -0,0 +1,303 @@
+			q.Add("showabnormaldetails", strconv.FormatBool(*opts.ShowAbnormalDetails))
+		}
+		if opts.EndDate != nil && !opts.EndDate.IsZero() {
+			q.Add("enddate", opts.EndDate.Format("01/02/2006"))
+		}
+		if opts.HideDuplicate != nil {
+			q.Add("hideduplicate", strconv.FormatBool(*opts.HideDuplicate))
+		}
+
+		if opts.Pagination != nil {
+			if opts.Pagination.Limit > 0 {
+				q.Add("limit", strconv.Itoa(opts.Pagination.Limit))
+			}
+
+			if opts.Pagination.Offset > 0 {
+				q.Add("offset", strconv.Itoa(opts.Pagination.Offset))
+			}
+		}
+	}
+
+	out := &listLabResultsResponse{}
+
+	_, err := h.Get(ctx, fmt.Sprintf("chart/%s/labresults", patientID), q, out)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ListLabResultsResult{
+		LabResults: out.LabResults,
+		Pagination: makePaginationResult(out.Next, out.Previous, out.TotalCount),
+	}, nil
+}
+
+type LabResultAttachmentType string
+
+const (
+	LabResultAttachmentTypeBMP  LabResultAttachmentType = "BMP"
+	LabResultAttachmentTypeGIF  LabResultAttachmentType = "GIF"
+	LabResultAttachmentTypeJPG  LabResultAttachmentType = "JPG"
+	LabResultAttachmentTypeJPEG LabResultAttachmentType = "JPEG"
+	LabResultAttachmentTypePDF  LabResultAttachmentType = "PDF"
+	LabResultAttachmentTypeTIF  LabResultAttachmentType = "TIF"
+	LabResultAttachmentTypeTIFF LabResultAttachmentType = "TIFF"
+)
+
+type AddLabResultDocumentOptions struct {
+	// AttachmentContents must be Base64 encoded
+	AttachmentContents io.Reader               `json:"attachmentcontents"`
+	AttachmentType     LabResultAttachmentType `json:"attachmenttype"`
+	InternalNote       *string                 `json:"internalnote"`
+	NoteToPatient      *string                 `json:"notetopatient"`
+	ObservationDate    *time.Time              `json:"observationdate"`
+	OriginalFilename   *string                 `json:"originalfilename"`
+	// 1 = high, 2 = normal
+	Priority    *string `json:"priority"`
+	ResultNotes *string `json:"resultnotes"`
+	// Final, Partial, Pending, Preliminary, Corrected, Cancelled
+	ResultStatus *string `json:"resultstatus"`
+	TieToOrderID *int    `json:"tietoorderid"`
+}
+
+type addLabResultDocumentResponse struct {
+	ErrorMessage string `json:"errormessage"`
+	LabResultID  int    `json:"labresultid"`
+	Success      bool   `json:"success"`
+}
+
+// AddLabResultDocument creates a lab result document record of a specific patient
+//
+// POST /v1/{practiceid}/patients/{patientid}/documents/labresult
+//
+// https://docs.athenahealth.com/api/api-ref/document-type-lab-result#Add-lab-result-document-to-patient's-chart
+func (h *HTTPClient) AddLabResultDocumentReader(ctx context.Context, patientID string, departmentID string, opts *AddLabResultDocumentOptions) (int, error) {
+	form := NewFormURLEncoder()
+
+	form.AddString("departmentid", departmentID)
+
+	if opts != nil {
+		if opts.AttachmentContents != nil {
+			form.AddReader("attachmentcontents", newBase64Reader(opts.AttachmentContents))
+		}
+		if len(opts.AttachmentType) > 0 {
```

Comment:
We may want to return an error if the attachment type is empty. Might be a good idea to capture any/all client-side errors and join. Checkout my comment here https://github.com/eleanorhealth/go-athenahealth/pull/111#discussion_r1293486940

Reply:
Addressed in above comments